core/iwasm/interpreter/wasm_loader.c

```diff
@@ -4836,7 +4836,18 @@ load_data_segment_section(const uint8 *buf, const uint8 *buf_end,
 
             dataseg->data_length = data_seg_len;
             CHECK_BUF(p, p_end, data_seg_len);
-            dataseg->data = (uint8 *)p;
+            if (!is_load_from_file_buf) {
+                if (!(dataseg->data = loader_malloc(
+                          dataseg->data_length, error_buf, error_buf_size))) {
+                    return false;
+                }
```

Comment:
> BTW, should also change calling other functions like load_user_section in load_from_sections, somewhat like below (to be checked):

Why do we need that? Shouldn't `is_load_from_file_buf` be enough since now we split `is_load_from_file_buf` and `wasm_binary_freeable` in `load_from_sections`? It would be the same as it was before #3389

Reply:
Thanks for the explanation, that's right. I pushed the change but we should probably rename `is_load_from_file_buf ` (in `wasm_const_str_list_insert`) since now `is_load_from_file_buf && !wasm_binary_freeable` gets passed to it.